### ISSUE TITLE
[AOTInductor] Seeded fallback nonce collision causes cross-rank reuse of device-pinned AOT cache entries

### ISSUE THREAD

**Issue by mengph** [pytorch/pytorch#191323]
### Summary

In AOT precompile mode, when AOTAutograd fails to generate a regular cache
key, `autograd_cache_key()` falls back to:

```python
return str(random.random()), []
```

This fallback depends on Python's application-global RNG state. If multiple
distributed ranks use the same Python seed and share
`TORCHINDUCTOR_CACHE_DIR`, they can generate the same fallback key.

In the reproducer below:

1. Rank 0 and rank 1 each capture a device-local CUDA Stream.
2. Dynamo creates a `get_external_object_by_index` node for the Stream.
3. The Stream is only used by an `is None` check, which Dynamo folds away,
   leaving the external-object node with no users.
4. AOTAutograd `check_cacheable()` still checks this dead node and raises
   `BypassAOTAutogradCache`.
5. AOT precompile catches the exception and calls `random.random()`.
6. Both ranks use `random.seed(0)`, so both fallback keys are
   `0.8444218515250481`.
7. Rank 0 writes an AOT cache entry compiled for CUDA:0.
8. Rank 1 uses the same fallback key and loads the CUDA:0 entry while running
   with a CUDA:1 input.
9. The reused wrapper creates indices on CUDA:0 and passes them together with
   a CUDA:1 embedding weight, causing a device mismatch.

The current implementation can be seen in
`torch/_functorch/_aot_autograd/autograd_cache.py`:

```python
except Exception:
    if torch._functorch.config.bypass_autograd_cache_key:
        ...
        return str(random.random()), []
```

### Expected behavior

A fallback nonce used as an AOT cache key should be unique across processes
even when applications intentionally initialize their Python RNGs with the
same seed.

Rank 1 should not load an AOT cache entry generated for rank 0 merely because
both processes have identical application RNG state.

Possible approaches include using an entropy source independent of the
application RNG, such as `uuid.uuid4()` or `secrets.token_hex()`.

Separately, a pure dead `get_external_object_by_index` node arguably should
not prevent regular AOT cache-key generation. A live external-object node
should remain non-cacheable unless it has stable hashing and reconstruction
semantics.

### Actual behavior

Both ranks generate the same fallback key:

```text
rank fallback keys: ['0.8444218515250481', '0.8444218515250481']
```

Both therefore address the same AOT cache directory:

```text
/tmp/aot_nonce_repro_.../aotautograd/0.8444218515250481
/tmp/aot_nonce_repro_.../aotautograd/0.8444218515250481
```

Rank 1 then reuses the rank 0 wrapper and fails with:

```text
Expected all tensors to be on the same device,
but got index is on cuda:0, different from other tensors on cuda:1
```

### Ablations

- Removing the captured Stream avoids the
  `get_external_object_by_index` cacheability failure.
- Using different Python seeds avoids this particular fallback-key collision.
- Giving each rank a separate `TORCHINDUCTOR_CACHE_DIR` avoids cross-rank
  reuse.
- The custom external embedding operator in the reproducer is not required
  to trigger the cache collision. It only prevents Inductor from fusing
  embedding into a Triton kernel, so the incorrect cache reuse produces a
  descriptive tensor device-mismatch error instead of a lower-level
  `CUDA driver error: invalid argument`.

### Minimal reproducer

Requirements:

- A PyTorch build exposing the experimental `aot_compile` API
- At least two CUDA devices

Run with:

```bash
CUDA_VISIBLE_DEVICES=0,1 python repro.py
```

```python
#!/usr/bin/env python3

import os
import random
import tempfile
from pathlib import Path

import torch
import torch.distributed as dist
import torch.multiprocessing as mp


WORLD_SIZE = 2
SEED = 0

# Keep embedding as an external call so that loading a wrapper generated for
# another device produces an explicit tensor device-mismatch error.
_LIB = torch.library.Library("nonce_repro", "DEF")
_LIB.define("embedding(Tensor weight, Tensor indices) -> Tensor")
_LIB.impl(
    "embedding",
    lambda weight, indices: torch.ops.aten.embedding.default(weight, indices),
    "CUDA",
)
_LIB.impl(
    "embedding",
    lambda weight, indices: weight.new_empty(
        (indices.numel(), weight.shape[1])
    ),
    "Meta",
)


class Model(torch.nn.Module):
    def __init__(self, stream):
        super().__init__()
        self.stream = stream

    def forward(self, weight):
        # Dynamo folds this condition but leaves a dead external-object node.
        if self.stream is None:
            raise AssertionError("unreachable")

        indices = torch.arange(4, device=weight.device)
        return torch.ops.nonce_repro.embedding.default(weight, indices)


class TraceFallbackRandom:
    """Expose calls to random.random() from autograd_cache_key's except path."""

    def __init__(self, rank):
        self.rank = rank
        self.value = None

    def random(self):
        value = random.random()
        self.value = str(value)
        print(
            f"[rank {self.rank}] "
            f"AOT except -> random.random() = {self.value}",
            flush=True,
        )
        return value


def worker(rank, cache_dir, rendezvous):
    from torch._dynamo import config as dynamo_config
    from torch._functorch._aot_autograd import autograd_cache

    torch.cuda.set_device(rank)
    dist.init_process_group(
        "gloo",
        init_method=f"file://{rendezvous}",
        rank=rank,
        world_size=WORLD_SIZE,
    )
    dynamo_config.enable_aot_compile = True

    # Only expose the cacheability exception; preserve the original behavior.
    original_check = autograd_cache.check_cacheable

    def check_cacheable(gm):
        try:
            return original_check(gm)
        except Exception as error:
            print(
                f"[rank {rank}] check_cacheable except: {error}",
                flush=True,
            )
            raise

    autograd_cache.check_cacheable = check_cacheable

    # Replace only the random object used by autograd_cache.py. The returned
    # value still comes from Python's real random.random().
    traced_random = TraceFallbackRandom(rank)
    autograd_cache.random = traced_random

    random.seed(SEED)
    weight = torch.randn(8, 4, device=f"cuda:{rank}")
    model = Model(torch.cuda.Stream(device=rank))

    if rank == 0:
        compiled = torch.compile(
            model,
            fullgraph=True,
        ).forward.aot_compile(((weight,), {}))

        compiled(model, weight)
        torch.cuda.synchronize()
        print("[rank 0] wrote the CUDA:0 AOT cache entry", flush=True)

    # Ensure rank 0 saves its entry before rank 1 performs its lookup.
    dist.barrier()

    if rank == 1:
        compiled = torch.compile(
            model,
            fullgraph=True,
        ).forward.aot_compile(((weight,), {}))

        try:
            compiled(model, weight)
            torch.cuda.synchronize()
        except RuntimeError as error:
            message = str(error)
            assert "same device" in message, message
            print(f"[rank 1] device mismatch: {message}", flush=True)
        else:
            raise AssertionError(
                "expected rank 1 to reuse the CUDA:0 wrapper"
            )

    dist.barrier()

    keys = [None] * WORLD_SIZE
    dist.all_gather_object(keys, traced_random.value)

    if rank == 0:
        key_dirs = [
            Path(cache_dir, "aotautograd", key)
            for key in keys
        ]

        assert keys[0] == keys[1]
        assert all(path.is_dir() for path in key_dirs)

        print(f"\nrank fallback keys: {keys}")
        print("same tmp AOT directory:")
        for path in key_dirs:
            print(path)

        print(
            "\nREPRODUCED: both ranks used the same "
            "random fallback directory."
        )

    dist.destroy_process_group()


def main():
    if not torch.cuda.is_available() or torch.cuda.device_count() < WORLD_SIZE:
        raise RuntimeError("this reproducer requires two CUDA devices")

    cache_dir = Path(tempfile.mkdtemp(prefix="aot_nonce_repro_"))
    os.environ["TORCHINDUCTOR_CACHE_DIR"] = str(cache_dir)
    os.environ["TORCHINDUCTOR_FX_GRAPH_CACHE"] = "1"

    rendezvous = cache_dir / "dist_rendezvous"

    print(f"shared tmp cache: {cache_dir}", flush=True)
    mp.spawn(
        worker,
        args=(str(cache_dir), str(rendezvous)),
        nprocs=WORLD_SIZE,
        join=True,
    )


if __name__ == "__main__":
    main()
```

Error logs
The reproducer intentionally catches the final `RuntimeError` so that it can
assert that the error is specifically a device mismatch.

```text
shared tmp cache: /tmp/aot_nonce_repro_ouj48ud9

[rank 0] check_cacheable except:
Unsupported call_function target
<function get_external_object_by_index at 0x...>.

Function module: torch._dynamo.graph_bytecode_inputs,
Function name: get_external_object_by_index

[rank 0] AOT except -> random.random() = 0.8444218515250481
[rank 0] wrote the CUDA:0 AOT cache entry

[rank 1] check_cacheable except:
Unsupported call_function target
<function get_external_object_by_index at 0x...>.

Function module: torch._dynamo.graph_bytecode_inputs,
Function name: get_external_object_by_index

[rank 1] AOT except -> random.random() = 0.8444218515250481

[rank 1] device mismatch:
Expected all tensors to be on the same device,
but got index is on cuda:0, different from other tensors on cuda:1
(when checking argument in method wrapper_CUDA__index_select)

rank fallback keys: ['0.8444218515250481', '0.8444218515250481']

same tmp AOT directory:
/tmp/aot_nonce_repro_ouj48ud9/aotautograd/0.8444218515250481
/tmp/aot_nonce_repro_ouj48ud9/aotautograd/0.8444218515250481

REPRODUCED: both ranks used the same random fallback directory.
```

### Versions

PyTorch version: 2.12.0
Installed torch package:
2.12.0a0+5aff3928d8.nv26.5.50603568

Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
CMake version: version 3.31.6
Libc version: glibc-2.39

Python version:
3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0]
Python platform:
Linux-6.2.0-26-generic-x86_64-with-glibc2.39

Is CUDA available: True
CUDA runtime version: 13.2.78
CUDA_MODULE_LOADING set to: LAZY

GPU models and configuration:
GPU 0: NVIDIA L20
GPU 1: NVIDIA L20
GPU 2: NVIDIA L20
GPU 3: NVIDIA L20
GPU 4: NVIDIA L20
GPU 5: NVIDIA L20
GPU 6: NVIDIA L20
GPU 7: NVIDIA L20

Nvidia driver version: 580.95.05

Relevant Python packages:
numpy==2.1.0
torch==2.12.0a0+5aff3928d8.nv26.5.50603568
torch_tensorrt==2.12.0a0
torchvision==0.27.0a0+6214cb20.nv26.5.50603568
triton==3.7.0+gitb4e20bb.nv26.5
triton_kernels==1.0.0+gitb4e20bb.nv26.5

cc @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste
